### PR TITLE
HPCC-29585 Improve compression of sequential options

### PR DESCRIPTION
**Pull request by ghalliday**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: system/jhtree/jhinplace.cpp [hunk 1]
```diff
@@ -53,6 +53,9 @@ static constexpr byte NSFcompressTrailing   = 0x20;     // compressed payload ha
 static constexpr byte NSFscaleN             = 0x40;     // filepositions are scaled by size N - useful for filepositions in fixed size files
 static constexpr byte NSFscaleFilepos       = 0x80;     // filepositions are scaled by nodesize
 
+static constexpr byte OFsequential          = 0x40;     // Options are a sequential range of values e.g. '0','1','2','3','4','5'.  Only first value stored
```
Comment (richardkchapman) [inline system/jhtree/jhinplace.cpp:56]: Should clarify that only first non-space value is stored?

File: system/jhtree/jhinplace.cpp [hunk 2]
```diff
@@ -567,17 +603,28 @@ void PartialMatch::serialize(MemoryBuffer & out)
         {
             byte offsetBytes = bytesRequired(getMaxOffset());
             byte countBytes = bytesRequired(getCount()-1);
+            byte sequentialFlags = getSequentialOptionFlags();
 
-            byte sizeInfo = offsetBytes;
+            byte sizeInfo = sequentialFlags | offsetBytes;
```
Comment (richardkchapman) [inline system/jhtree/jhinplace.cpp:608]: What is max value of offsetBytes? Do we know never going to collide with flags?

File: system/jhtree/jhinplace.cpp [hunk 3]
```diff
@@ -750,12 +797,27 @@ bool PartialMatch::squash()
     return dirty;
 }
 
-const byte * PartialMatch::queryNullRow()
+const byte * PartialMatch::queryNullRow() const
 {
     return builder->queryNullRow();
 }
 
 //MORE: Pass this in...
```
Comment (richardkchapman) [inline system/jhtree/jhinplace.cpp:805]: Is this comment relevant to the function you just added after it?

**Review by richardkchapman — COMMENTED** [review]
Some questions. Code that tests the flag looks right, but can you be sure you found all places that you need to test the flag?

File: system/jhtree/jhinplace.cpp (hunk 2 above)
Comment (ghalliday) [reply system/jhtree/jhinplace.cpp:607]: It will never collide.
Offset bytes will be <= 2 for our key files.  Maximum supported value is 7 (to allow 4 byte offsets).

File: system/jhtree/jhinplace.cpp (hunk 3 above)
Comment (ghalliday) [reply system/jhtree/jhinplace.cpp:804]: I don't think it applies to either function - I will delete it

**Comment by ghalliday** [comment]
@mckellyln rebased onto merged code.  I don't think you probably looked at this as part of #17370.

File: system/jhtree/jhinplace.cpp (hunk 1 above)
Comment (ghalliday) [reply system/jhtree/jhinplace.cpp:55]: clarified

**Review by mckellyln — APPROVED** [review]
Approved
